### PR TITLE
Rewrite KThread to be more accurate

### PR DESCRIPTION
**Pull request by bunnei**
This change rewrites our KThread implementation to more closely match the Switch's real kernel.

### CODE REVIEW

File: src/core/hle/kernel/k_thread_queue.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+#pragma once
+
+#include "core/hle/kernel/k_thread.h"
+
+namespace Kernel {
+
+class KThreadQueue {
+public:
+    KThreadQueue(KernelCore& kernel) : kernel{kernel}, wait_list{} {}
```
Comment (lioncash) [inline src/core/hle/kernel/k_thread_queue.h:13]:
```suggestion
    explicit KThreadQueue(KernelCore& kernel) : kernel{kernel} {}
```

File: src/core/hle/kernel/k_thread_queue.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            thread->SetState(ThreadState::Runnable);
+            thread->SetSleepingQueue(nullptr);
+
+            return thread;
+        }
+    }
+
+private:
+    KernelCore& kernel;
+    KThread::WaiterList wait_list;
```
Comment (lioncash) [inline src/core/hle/kernel/k_thread_queue.h:78]:
```suggestion
    KThread::WaiterList wait_list{};
```

File: src/core/hle/kernel/session.h [hunk 3]
```diff
@@ -39,6 +39,8 @@ class Session final : public KSynchronizationObject {
 
     bool IsSignaled() const override;
 
+    void Finalize() override{};
```
Comment (lioncash) [inline src/core/hle/kernel/session.h:42]:
```suggestion
    void Finalize() override {}
```

File: src/common/common_funcs.h [hunk 4]
```diff
@@ -101,6 +101,21 @@ __declspec(dllimport) void __stdcall DebugBreak(void);
         }                                                                                          \
     }
 
+#define R_SUCCEEDED(res) (res.IsSuccess())
```
Comment (lioncash) [inline src/common/common_funcs.h:104]: Are the `R_` macros planned to be removed after the migration of the majority of the kernel?

File: src/common/common_funcs.h [hunk 5]
```diff
@@ -101,6 +101,21 @@ __declspec(dllimport) void __stdcall DebugBreak(void);
...
+#define R_FAILED(res) (res.IsError())
```
Comment (lioncash) [inline src/common/common_funcs.h:105]: This seems to only exist for the R_TRY statement below. Maybe just replace it with the expression?

File: src/common/common_funcs.h (hunk 4 above)
Comment (bunnei) [reply src/common/common_funcs.h:104]: Yes, just trying to keep migration efforts a little more straightforward for now -- makes comparing a bit easier

**Comment by v1993** [comment]
Having trouble building this:

```
In file included from ../src/./core/hardware_properties.h:10,
                 from ../src/./core/hle/kernel/global_scheduler_context.h:12,
                 from ../src/./core/hle/kernel/k_scheduler.h:14,
                 from ../src/core/hle/kernel/k_light_lock.cpp:6:
../src/./common/bit_util.h: In function ‘constexpr u32 Common::MostSignificantBit32(u32)’:
../src/./common/bit_util.h:21:40: error: ‘countl_zero’ is not a member of ‘std’
   21 |     return 31U - static_cast<u32>(std::countl_zero(value));
      |                                        ^~~~~~~~~~~
../src/./common/bit_util.h: In function ‘constexpr u32 Common::MostSignificantBit64(u64)’:
../src/./common/bit_util.h:25:40: error: ‘countl_zero’ is not a member of ‘std’
   25 |     return 63U - static_cast<u32>(std::countl_zero(value));
      |                                        ^~~~~~~~~~~
```

Gcc 10 is not new enough?

File: src/core/hle/kernel/k_scheduler_lock.h [hunk 6]
```diff
@@ -22,54 +23,53 @@ class KAbstractSchedulerLock {
     explicit KAbstractSchedulerLock(KernelCore& kernel_) : kernel{kernel_} {}
 
     bool IsLockedByCurrentThread() const {
-        return this->owner_thread == kernel.GetCurrentEmuThreadID();
+        return this->owner_thread == GetCurrentThreadPointer(kernel);
```
Comment (ogniK5377) [inline src/core/hle/kernel/k_scheduler_lock.h:26]:
```suggestion
        return owner_thread == GetCurrentThreadPointer(kernel);
```

File: src/core/hle/kernel/k_scheduler_lock.h [hunk 7]
```diff
@@ -22,54 +23,53 @@ class KAbstractSchedulerLock {
...
     }
 
     void Lock() {
         if (this->IsLockedByCurrentThread()) {
```
Comment (ogniK5377) [inline src/core/hle/kernel/k_scheduler_lock.h:30]:
```suggestion
        if (IsLockedByCurrentThread()) {
```

File: src/core/hle/kernel/k_scheduler_lock.h [hunk 8]
```diff
@@ -22,54 +23,53 @@ class KAbstractSchedulerLock {
...
             // Increment count, take ownership.
-            this->lock_count = 1;
-            this->owner_thread = kernel.GetCurrentEmuThreadID();
+            lock_count = 1;
+            owner_thread = GetCurrentThreadPointer(kernel);
         }
     }
 
     void Unlock() {
         ASSERT(this->IsLockedByCurrentThread());
```
Comment (ogniK5377) [inline src/core/hle/kernel/k_scheduler_lock.h:50]:
```suggestion
        ASSERT(IsLockedByCurrentThread());
```

File: src/core/hle/kernel/k_thread.cpp [hunk 9]
```diff
@@ -0,0 +1,1048 @@
...
+        break;
+    default:
+        UNREACHABLE_MSG("KThread::Initialize: Unknown ThreadType {}", static_cast<u32>(type));
+        break;
+    }
+
+    // Set the ideal core ID and affinity mask.
+    virtual_ideal_core_id = virt_core;
+    physical_ideal_core_id = phys_core;
+    virtual_affinity_mask = (static_cast<u64>(1) << virt_core);
```
Comment (ogniK5377) [inline src/core/hle/kernel/k_thread.cpp:108]:
```suggestion
    virtual_affinity_mask = (1ULL << virt_core);
```

File: src/core/hle/kernel/k_thread.cpp [hunk 10]
```diff
@@ -0,0 +1,1048 @@
...
+void KThread::Unpin() {
+    ASSERT(kernel.GlobalSchedulerContext().IsLocked());
+
+    // Set ourselves as unpinned.
+    GetStackParameters().is_pinned = false;
+
+    // Enable core migration.
+    ASSERT(num_core_migration_disables == 1);
+    {
+        --num_core_migration_disables;
```
Comment (ogniK5377) [inline src/core/hle/kernel/k_thread.cpp:356]:
```suggestion
        num_core_migration_disables--;
```

File: src/core/hle/kernel/k_thread.cpp [hunk 11]
```diff
@@ -0,0 +1,1048 @@
...
+
+            // Resume.
+            Resume(SuspendType::Thread);
+        }
+    }
+
+    // If the thread is now paused, update the pinned waiter list.
+    if (activity == Svc::ThreadActivity::Paused) {
+        bool thread_is_pinned = false;
+        bool thread_is_current;
```
Comment (ogniK5377) [inline src/core/hle/kernel/k_thread.cpp:687]:
```suggestion
        bool thread_is_current{};
```

File: src/core/hle/kernel/process.cpp [hunk 12]
```diff
@@ -137,6 +136,23 @@ std::shared_ptr<ResourceLimit> Process::GetResourceLimit() const {
     return resource_limit;
 }
 
+void Process::IncrementThreadCount() {
+    ASSERT(num_threads >= 0);
+    ++num_created_threads;
```
Comment (ogniK5377) [inline src/core/hle/kernel/process.cpp:141]:
```suggestion
    num_created_threads++;
```

**Review by ogniK5377 — APPROVED** [review]
Few tiny nits

**Comment by lat9nq** [comment]
This is already known: just documenting it here. 2136c2c86c9f279f94b717b4a6963a97fe4b218d and 8ba88277bcdd68d29d1311f7d8a9b9f5a36cb239 causes softlocks in Linux while booting Super Smash Bros. Ultimate and Boomerang Fu.

File: src/core/hle/kernel/k_scheduler_lock.h (hunk 6 above)
Comment (bunnei) [reply src/core/hle/kernel/k_scheduler_lock.h:26]: Resolve already in a later commit

**Comment by IGCIT** [comment]
random softlocks testing this, env: ArchLinux, pokemon sword update 1.2.1

softlock entering dynamax raid
```
[1833.497952] Debug <Critical> core/hle/kernel/svc.cpp:CreateThread:1453: Assertion Failed!
[1833.500336] Debug <Critical> core/hle/kernel/svc.cpp:CreateThread:1453: Assertion Failed!
[1833.500612] Debug <Critical> core/hle/kernel/svc.cpp:CreateThread:1453: Assertion Failed!
[1833.500759] Debug <Critical> core/hle/kernel/svc.cpp:CreateThread:1453: Assertion Failed!
[1834.570516] Debug.Emulated <Critical> core/hle/kernel/svc.cpp:Break:574: Signalling debugger, PANIC! info1=0x0000001024864BF0, info2=0x0000000000000004
[1834.570520] Debug.Emulated <Critical> core/hle/kernel/svc.cpp:operator():556: debug_buffer_err_code=2F6202
[1834.570523] Debug.Emulated <Critical> core/hle/kernel/svc.cpp:Break:623: Emulated program broke execution! reason=0x0000000000000000, info1=0x0000001024864BF0, info2=0x0000000000000004
[1834.570525] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:266: Backtrace, sp=000000000000000E, pc=000000000BC57BA8
[1834.570528] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:267: Module Name         Address             Original Address    Offset              Symbol
[1834.570529] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:269: 
[1834.591798] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: sdk                 00000071001801E8    000000000BA591E8    00000000001801E8    _ZN2nn4diag6detail5AbortEPKNS_6ResultE
[1834.591804] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: sdk                 000000710017ED38    000000000BA57D38    000000000017ED38    _ZN2nn4diag6detail10VAbortImplEPKcS3_S3_iPKNS_6ResultEPKNS_2os17UserExceptionInfoERKNS0_14MetaLogContextES3_St9__va_list
[1834.591805] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: sdk                 000000710017EE04    000000000BA57E04    000000000017EE04    _ZN2nn4diag6detail9AbortImplEPKcS3_S3_iPKNS_6ResultES3_z
[1834.591806] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: sdk                 00000071001F291C    000000000BACB91C    00000000001F291C    _ZN2nn2fs8OpenFileEPNS0_10FileHandleEPKci
[1834.591807] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005E6388    00000000085EA388    00000000005E6388    
[1834.591807] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005E2974    00000000085E6974    00000000005E2974    
[1834.591808] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005DE390    00000000085E2390    00000000005DE390    
[1834.591815] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005E6F10    00000000085EAF10    00000000005E6F10    
[1834.591816] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005DE1D8    00000000085E21D8    00000000005DE1D8    
[1834.591816] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005BB390    00000000085BF390    00000000005BB390    
[1834.591817] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005BA7CC    00000000085BE7CC    00000000005BA7CC    
[1834.591818] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005BEBC8    00000000085C2BC8    00000000005BEBC8    
[1834.591819] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005BE948    00000000085C2948    00000000005BE948    
[1834.591819] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005E73D4    00000000085EB3D4    00000000005E73D4    
[1834.591820] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: main                00000071005D11D4    00000000085D51D4    00000000005D11D4    
[1834.591821] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: sdk                 0000007100329E84    000000000BC02E84    0000000000329E84    
[1834.591821] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: sdk                 000000710032E290    000000000BC07290    000000000032E290    
[1834.591822] Core.ARM <Error> core/arm/arm_interface.cpp:LogBacktrace:273: sdk                 00000070F4726FFC    FFFFFFFFFFFFFFFC    FFFFFFFFF4726FFC
```


softlock trainer battle start
```
[ 352.967403] Debug <Critical> core/hle/kernel/svc.cpp:CreateThread:1453: Assertion Failed!
[ 352.967570] Debug <Critical> core/hle/kernel/svc.cpp:CreateThread:1453: Assertion Failed!
[ 352.967665] Debug <Critical> core/hle/kernel/svc.cpp:CreateThread:1453: Assertion Failed!
[ 353.109432] Render.OpenGL <Info> video_core/renderer_opengl/gl_shader_cache.cpp:BuildShader:167: FS0F34B08BFDD37466
[ 353.110212] Render.OpenGL <Info> video_core/renderer_opengl/gl_shader_cache.cpp:BuildShader:167: FS0F34B08BFDD37466
[ 353.139788] Render.OpenGL <Info> video_core/renderer_opengl/gl_shader_cache.cpp:BuildShader:167: FS90EC8FFECB8331F8
[ 353.142616] Render.OpenGL <Info> video_core/renderer_opengl/gl_shader_cache.cpp:BuildShader:167: FS90EC8FFECB8331F8
[ 353.166164] Kernel <Critical> core/hle/kernel/k_address_arbiter.cpp:SignalAndModifyByWaitingCountIfEqual:214: Failed with error 64001
```

**Pull request merged**